Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/DirectoryInitializationService.java

```diff
@@ -69,21 +72,55 @@ else if (PermissionsHandler.hasWriteAccess(this))
+    private void initializeInternalStorage()
     {
-        String BaseDir = NativeLibrary.GetUserDirectory();
-        String ConfigDir = BaseDir + File.separator + "Config";
-
-        // Copy assets if needed
-        NativeLibrary.CreateUserFolders();
-        copyAssetFolder("GC", BaseDir + File.separator + "GC", false);
-        copyAssetFolder("Shaders", BaseDir + File.separator + "Shaders", false);
-        copyAssetFolder("Wii", BaseDir + File.separator + "Wii", false);
-
-        // Always copy over the GCPad config in case of change or corruption.
-        // Not a user configurable file.
-        copyAsset("GCPadNew.ini", ConfigDir + File.separator + "GCPadNew.ini", true);
-        copyAsset("WiimoteNew.ini", ConfigDir + File.separator + "WiimoteNew.ini", false);
+        File sysDirectory = new File(getFilesDir(), "Sys");
+
+        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
+        String revision = NativeLibrary.GetGitRevision();
+        if (!preferences.getString("sysDirectoryVersion", "").equals(revision))
+        {
+            // There is no extracted Sys directory, or there is a Sys directory from another
+            // version of Dolphin that might contain outdated files. Let's (re-)extract Sys.
+            deleteDirectoryRecursively(sysDirectory);
+            copyAssetFolder("Sys", sysDirectory, true);
+
+            SharedPreferences.Editor editor = preferences.edit();
+            editor.putString("sysDirectoryVersion", revision);
+            editor.apply();
+        }
+
+        // Let the native code know where the Sys directory is.
+        SetSysDirectory(sysDirectory.getPath());
+    }
+
+    private void initializeExternalStorage()
+    {
+        // Create User directory structure and copy some NAND files from the extracted Sys directory.
+        CreateUserDirectories();
+
+        // These two files must contain specific values in order for controller input to work
+        // as intended (they aren't user configurable), so we overwrite them in case the user
+        // has modified them manually.
+        //
+        // ...except WiimoteNew.ini contains the user configurable settings for Wii Remote
+        // extensions in addition to all of its lines that aren't user configurable, so in order
+        // to not lose the user's extension selections, we don't overwrite that file if it exists.
+        //
+        // TODO: Redo the Android controller system entirely. The Config directory should store
+        // the user's controller configuration, not special values that the user mustn't touch.
+        String configDirectory = NativeLibrary.GetUserDirectory() + File.separator + "Config";
+        copyAsset("GCPadNew.ini", new File(configDirectory, "GCPadNew.ini"), true);
+        copyAsset("WiimoteNew.ini", new File(configDirectory,"WiimoteNew.ini"), false);
+    }
+
+    private static void deleteDirectoryRecursively(File file)
```

Comment:
This method does nothing; it recurses but never deletes anything.

Reply:
Thanks for catching that! I must've missed that line when rebasing. (The whole file had changed indentation style in master, so nothing could be merged automatically...)